aten/src/ATen/native/xnnpack/MaxPooling.cpp

```diff
@@ -0,0 +1,367 @@
+         // Input
+  return (4 == input.dim()) &&
+         // (c10::DeviceType::CPU == input.device().type()) &&
+         (at::Backend::CPU == input.options().backend()) &&
+         (kFloat == input.scalar_type()) &&
+         // Output
+         (shape(input.size(Layout::Activation4D::height), height) > 0) &&
+         (shape(input.size(Layout::Activation4D::width), width) > 0) &&
+         true;
+}
+
+Tensor run(
+    const Context& context,
+    const Tensor& input) {
+  using namespace internal;
+
+  const Tensor input_nhwc = input.contiguous(MemoryFormat::ChannelsLast);
+  const Tensor padded_input_nhwc = allocate_padded_if_needed(input_nhwc);
+
+  TORCH_CHECK(
+      usable(padded_input_nhwc, context.output.height, context.output.width),
+      "XNNPACK MaxPool2d not usable! "
+      "Reason: The provided input tensor is either invalid or unsupported by XNNPACK.");
+
+  Tensor output_nhwc = empty_with_tail_padding(
+      {
+        padded_input_nhwc.size(Layout::Activation4D::batch),
+        context.output.channels,
+        shape(padded_input_nhwc.size(Layout::Activation4D::height), context.output.height),
+        shape(padded_input_nhwc.size(Layout::Activation4D::width), context.output.width),
+      },
+      padded_input_nhwc.options().dtype(),
+      MemoryFormat::ChannelsLast,
+      padded_input_nhwc.names());
+
+  const xnn_status setup_status = xnn_setup_max_pooling2d_nhwc_f32(
+      context.max_pool_op.get(),                            // operator
+      padded_input_nhwc.size(Layout::Activation4D::batch),  // batch_size
+      padded_input_nhwc.size(Layout::Activation4D::height), // input_height
+      padded_input_nhwc.size(Layout::Activation4D::width),  // input_width
+      padded_input_nhwc.data_ptr<float>(),                  // input
+      output_nhwc.data_ptr<float>(),                        // output
+      caffe2::xnnpack_threadpool());                        // threadpool
+
+  TORCH_CHECK(
+      xnn_status_success == setup_status,
+      "xnn_setup_max_pooling2d_nhwc_f32 failed!");
+
+  const xnn_status run_status = xnn_run_operator(
+      context.max_pool_op.get(),      // operator
+      caffe2::xnnpack_threadpool());  // threadpool
+
+  TORCH_INTERNAL_ASSERT(
+      xnn_status_success == run_status,
+      "xnn_run_operator failed!");
+
+  return output_nhwc.contiguous(input.suggest_memory_format());
+}
+
+Tensor create_and_run(
+    const Tensor& input,
+    const IntArrayRef kernel,
+    const IntArrayRef padding,
+    const IntArrayRef stride,
+    const IntArrayRef dilation,
+    const bool ceil_mode,
+    const float output_min,
+    const float output_max) {
+  using namespace internal;
+
+  return internal::max_pool2d::run(
+      internal::max_pool2d::create(
+          input.size(Layout::Activation4D::channels),
+          kernel,
+          padding,
+          stride,
+          dilation,
+          ceil_mode,
+          output_min,
+          output_max),
+      input);
+}
```

Comment:
This seems like an unnecessary layer of abstraction since we're not persisting the Context.  If you inline both create and run directly into `max_pool2d`, can you eliminate Context and shorten/simplify the entire diff?

Reply:
Don't we want to maintain the ability to separate create and run in the future?

Reply:
Why?  We don't expect it will ever have a significant perf improvement over this implementation, right?